### ISSUE TITLE
SQL generation hangs forever for joins with empty suffix

### ISSUE THREAD

**Issue by mgirlich** [tidyverse/dplyr#3220]
See the reprex

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(dbplyr)
#> 
#> Attaching package: 'dbplyr'
#> The following objects are masked from 'package:dplyr':
#> 
#>     ident, sql

data <- tibble(
  a = 1:3,
  b = letters[1:3]
)

# works
left_join(
  tbl_lazy(data, src = simulate_sqlite()),
  tbl_lazy(data, src = simulate_sqlite()),
  by = "a",
  suffix = c("_1", "_2")
) %>% 
  show_query()
#> <SQL> SELECT `TBL_LEFT`.`a` AS `a`, `TBL_LEFT`.`b` AS `b_1`, `TBL_RIGHT`.`b` AS `b_2`
#>   FROM `df` AS `TBL_LEFT`
#>   LEFT JOIN `df` AS `TBL_RIGHT`
#>   ON (`TBL_LEFT`.`a` = `TBL_RIGHT`.`a`)

# hangs forever
left_join(
  tbl_lazy(data, src = simulate_sqlite()),
  tbl_lazy(data, src = simulate_sqlite()),
  by = "a",
  suffix = c("", "_2")
)
```

<details><summary>Session info</summary>

``` r
devtools::session_info()
#> Warning in as.POSIXlt.POSIXct(Sys.time()): unknown timezone 'default/
#> Europe/Berlin'
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.0 (2017-04-21)
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  tz       <NA>                        
#>  date     2017-11-23
#> Packages -----------------------------------------------------------------
#>  package    * version    date       source                           
#>  assertthat   0.2.0      2017-04-11 CRAN (R 3.4.0)                   
#>  backports    1.1.1      2017-09-25 CRAN (R 3.4.2)                   
#>  base       * 3.4.0      2017-04-21 local                            
#>  bindr        0.1        2016-11-13 cran (@0.1)                      
#>  bindrcpp     0.2        2017-06-17 CRAN (R 3.4.0)                   
#>  compiler     3.4.0      2017-04-21 local                            
#>  datasets   * 3.4.0      2017-04-21 local                            
#>  DBI          0.7-12     2017-11-10 Github (rstats-db/DBI@6c111b7)   
#>  dbplyr     * 1.1.0.9000 2017-11-22 Github (tidyverse/dbplyr@a424f67)
#>  devtools     1.13.4     2017-11-09 CRAN (R 3.4.2)                   
#>  digest       0.6.12     2017-01-27 CRAN (R 3.4.0)                   
#>  dplyr      * 0.7.4.9000 2017-11-22 Github (tidyverse/dplyr@fc66342) 
#>  evaluate     0.10       2016-10-11 CRAN (R 3.4.0)                   
#>  glue         1.2.0      2017-10-29 CRAN (R 3.4.0)                   
#>  graphics   * 3.4.0      2017-04-21 local                            
#>  grDevices  * 3.4.0      2017-04-21 local                            
#>  htmltools    0.3.6      2017-04-28 CRAN (R 3.4.0)                   
#>  knitr        1.17       2017-08-10 CRAN (R 3.4.1)                   
#>  magrittr     1.5        2014-11-22 CRAN (R 3.4.0)                   
#>  memoise      1.1.0      2017-04-21 CRAN (R 3.4.0)                   
#>  methods    * 3.4.0      2017-04-21 local                            
#>  pillar       0.0.0.9000 2017-11-22 Github (r-lib/pillar@5a082e1)    
#>  pkgconfig    2.0.1      2017-03-21 cran (@2.0.1)                    
#>  purrr        0.2.4      2017-10-18 cran (@0.2.4)                    
#>  R6           2.2.2      2017-06-17 cran (@2.2.2)                    
#>  Rcpp         0.12.13    2017-09-28 cran (@0.12.13)                  
#>  rlang        0.1.4      2017-11-05 cran (@0.1.4)                    
#>  rmarkdown    1.8        2017-11-17 CRAN (R 3.4.2)                   
#>  rprojroot    1.2        2017-01-16 CRAN (R 3.4.0)                   
#>  stats      * 3.4.0      2017-04-21 local                            
#>  stringi      1.1.6      2017-11-17 CRAN (R 3.4.2)                   
#>  stringr      1.2.0      2017-02-18 CRAN (R 3.4.0)                   
#>  tibble       1.3.4.9003 2017-11-22 Github (tidyverse/tibble@60281b3)
#>  tidyselect   0.2.3      2017-11-06 CRAN (R 3.4.2)                   
#>  tools        3.4.0      2017-04-21 local                            
#>  utils      * 3.4.0      2017-04-21 local                            
#>  withr        2.1.0      2017-11-01 cran (@2.1.0)                    
#>  yaml         2.1.14     2016-11-12 CRAN (R 3.4.0)
```

</details>

**Comment by JohnMount** [comment]
Likely also causing [`sparklyr` issue 1130](https://github.com/rstudio/sparklyr/issues/1130).

**Comment by krlmlr** [comment]
Thanks, confirmed.

**Comment by hadley** [comment]
Slightly more minimal reprex:

``` r
library(dplyr, warn.conflicts = FALSE)
library(dbplyr, warn.conflicts = FALSE)

df <- lazy_frame(
  a = 1:3,
  b = letters[1:3],
  src = simulate_sqlite()
)

left_join(df, df, by = "a", suffix = c("_1", "_2")) %>% show_query()
#> <SQL> SELECT `TBL_LEFT`.`a` AS `a`, `TBL_LEFT`.`b` AS `b_1`, `TBL_RIGHT`.`b` AS `b_2`
#>   FROM `df` AS `TBL_LEFT`
#>   LEFT JOIN `df` AS `TBL_RIGHT`
#>   ON (`TBL_LEFT`.`a` = `TBL_RIGHT`.`a`)
#> <SQL> SELECT `TBL_LEFT`.`a` AS `a`, `TBL_LEFT`.`b` AS `b_1`, `TBL_RIGHT`.`b` AS `b_2`
#>   FROM `df` AS `TBL_LEFT`
#>   LEFT JOIN `df` AS `TBL_RIGHT`
#>   ON (`TBL_LEFT`.`a` = `TBL_RIGHT`.`a`)

# Hangs forever
# left_join(df, df, by = "a", suffix = c("", "_2")) %>% show_query()
```

If I ctrl + break the hang, I get the following `traceback()`:

```R
8: add_suffixes(x_names, y_names, suffix$x) at sql-query.R#96
7: join_vars(op_vars(x), op_vars(y), type = type, by = by, suffix = suffix) at tbl-lazy.R#228
6: add_op_join(x, y, "left", by = by, copy = copy, suffix = suffix, 
       auto_index = auto_index, ...) at tbl-sql.r#209
5: left_join.tbl_lazy(df, df, by = "a", suffix = c("", "_2"))
4: left_join(df, df, by = "a", suffix = c("", "_2"))
3: eval(lhs, parent, parent)
2: eval(lhs, parent, parent)
1: left_join(df, df, by = "a", suffix = c("", "_2")) %>% show_query()
```

**Comment by hadley** [comment]
And indeed the problem is this while:

```R
    while (nm %in% y || nm %in% out) {
      nm <- paste0(nm, suffix)
    }
```

which keeps adding suffixes until unique, which obviously isn't going to work here

**Comment by hadley** [comment]
Fixed in a85cd3702b133d4e428d6aad871812928460000c
